File: 351002/Zakrevskaya/src/marker/marker.service.ts
import { Injectable, NotFoundException, BadRequestException } from '@nestjs/common';
import { MarkerRepository } from './marker.repository';
import { CreateMarkerDto, MarkerResponseDto } from './dto/marker.dto';
import { Marker } from './marker.entity';

@Injectable()
export class MarkerService {
  constructor(private readonly markerRepository: MarkerRepository) {}

  async create(createMarkerDto: CreateMarkerDto): Promise<MarkerResponseDto> {
    this.validateMarkerDto(createMarkerDto);
    const marker = await this.markerRepository.create(createMarkerDto);
    return this.toResponseDto(marker);
  }

  async update(id: number, updateMarkerDto: CreateMarkerDto): Promise<MarkerResponseDto> {
    this.validateMarkerDto(updateMarkerDto);
    const marker = await this.markerRepository.update(id, updateMarkerDto);
    return this.toResponseDto(marker);
  }

  async delete(id: number): Promise<void> {
    const marker = await this.markerRepository.findById(id);
    if (!marker) throw new NotFoundException('Marker not found');
    await this.markerRepository.delete(id);
  }

  async findById(id: number): Promise<MarkerResponseDto> {
    const marker = await this.markerRepository.findById(id);
    if (!marker) throw new NotFoundException('Marker not found');
    return this.toResponseDto(marker);
  }

  async findAll(page: number = 1, limit: number = 10): Promise<MarkerResponseDto[]> {
    const [markers] = await this.markerRepository.findAll({ page, limit });
    return markers.map(marker => this.toResponseDto(marker));
  }

  private validateMarkerDto(dto: CreateMarkerDto) {
    if (dto.name.length < 2 || dto.name.length > 32) {
      throw new BadRequestException('Name must be 2-32 characters');
    }
  }

  private toResponseDto(marker: Marker): MarkerResponseDto {
    return {
      id: marker.id,
      name: marker.name,
    };
  }
}

File: 351004/Dashkevich/Redis_lab/publisher/src/main/java/com/adashkevich/redis/lab/service/MarkerService.java
package com.adashkevich.redis.lab.service;

import com.adashkevich.redis.lab.dto.request.MarkerRequestTo;
import com.adashkevich.redis.lab.dto.response.MarkerResponseTo;
import com.adashkevich.redis.lab.exception.ConflictException;
import com.adashkevich.redis.lab.exception.NotFoundException;
import com.adashkevich.redis.lab.model.News;
import com.adashkevich.redis.lab.repository.MarkerRepository;
import com.adashkevich.redis.lab.repository.NewsRepository;
import com.adashkevich.redis.lab.mapper.MarkerMapper;
import com.adashkevich.redis.lab.model.Marker;
import org.springframework.cache.annotation.CacheEvict;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@Transactional(readOnly = true)
public class MarkerService {
    private final MarkerRepository repo;
    private final NewsRepository newsRepo;
    private final MarkerMapper mapper;

    public MarkerService(MarkerRepository repo, NewsRepository newsRepo, MarkerMapper mapper) {
        this.repo = repo;
        this.newsRepo = newsRepo;
        this.mapper = mapper;
    }

    @CacheEvict(cacheNames = {"markers"}, allEntries = true)
    @Transactional
    public MarkerResponseTo create(MarkerRequestTo dto) {
        ensureNameUnique(dto.name, null);
        Marker saved = repo.save(mapper.toEntity(dto));
        return mapper.toResponse(saved);
    }

    @Cacheable(cacheNames = "markers")
    public List<MarkerResponseTo> getAll() {
        return repo.findAll(Sort.by("id")).stream().map(mapper::toResponse).toList();
    }

    @Cacheable(cacheNames = "marker", key = "#id")
    public MarkerResponseTo getById(Long id) {
        return mapper.toResponse(findExisting(id));
    }

    @CacheEvict(cacheNames = {"markers", "marker", "news", "newsItem"}, allEntries = true)
    @Transactional
    public MarkerResponseTo update(Long id, MarkerRequestTo dto) {
        Marker existing = findExisting(id);
        ensureNameUnique(dto.name, id);
        existing.setName(dto.name);
        return mapper.toResponse(repo.save(existing));
    }

    @CacheEvict(cacheNames = {"markers", "marker", "news", "newsItem"}, allEntries = true)
    @Transactional
    public void delete(Long id) {
        Marker marker = findExisting(id);
        for (News news : newsRepo.findAll()) news.getMarkers().removeIf(m -> id.equals(m.getId()));
        repo.delete(marker);
    }

    private Marker findExisting(Long id) {
        return repo.findById(id).orElseThrow(() -> new NotFoundException("Marker not found", "40430"));
    }

    private void ensureNameUnique(String name, Long selfId) {
        boolean exists = selfId == null ? repo.existsByNameIgnoreCase(name) : repo.existsByNameIgnoreCaseAndIdNot(name, selfId);
        if (exists) throw new ConflictException("Marker name must be unique", "40930");
    }
}
